### PR TITLE
Motion substream

### PR DESCRIPTION
**Pull request by tosiara**
This will create a secondary live view stream with a half reduced frame size
All other settings are inherited from the main live stream

Few notes
* Introduced new setting to specify a new port listener
* Backwards compatible with older config - default setting is 'off'
* Scale is hardcoded to be 50%
* If resized frame is not modulo of 8 - substream will be disabled
* There is also small optimization made to send frame to stream only if there are connected cleints
* Smaller MJPEG stream saves network bandwidth and motion's CPU

Smoke-tested on multpile x86 and ARM devices with netcam and V4L. No crashes or additional CPU overhead logged during past 24 hours. Seems to work stable

### CODE REVIEW

**Comment by tosiara** [comment]
> homepage should use the sub stream or main stream I think.

I can do that. If substream is enabled it will show it. If not - leave as is

> Can we get more flexibility in the scaling by using ffmpeg's scaling functionality (sws_scale() etc) as netcam_rtsp_resize does

Probably yes, I will take a look. Initially I was looking for a quick solution without cpu overhead and came to this "yuv420p scale by half" solution. Ideally, we would need to be able to specify any secondary resolution and do resize, not just scale.

**Comment by tosiara** [comment]
I have completed extended testing of substream and did not notice any issues like memory leak or crash. So should be good for initial release if we decide to merge it. Or we can keep it for the next major release. Or reject :)
Notes:

* Supports fixed scale by 50% of YUV420P
* Does not support GRAY pixel format
* Does not support resolutions that are not multiple of 16
* Backwards compatible: if secondary port is not specified in the config - the additional stream is not initialized

File: motion.c [hunk 1]
```diff
@@ -1141,6 +1142,29 @@ static int motion_init(struct context *cnt)
...
+    /* Initialize 50% scaled substream server if substream port is specified to not 0
+       But only if dimensions are 8-modulo after scaling. Otherwise disable substream */
+    if (cnt->conf.substream_port){
+        if ((cnt->conf.width / 2) % 8 == 0  && (cnt->conf.height / 2) % 8 == 0
+        && cnt->imgs.type == VIDEO_PALETTE_YUV420P){
+            if (stream_init (&(cnt->substream), cnt->conf.substream_port, cnt->conf.stream_localhost,
+                cnt->conf.ipv6_enabled) == -1) {
+                MOTION_LOG(ERR, TYPE_ALL, SHOW_ERRNO, "Problem enabling motion-substream server in port %d",
+                           cnt->conf.substream_port);
+                cnt->finish = 1;
```
Comment (Mr-DaveDev) [inline motion.c]: I think we will need a `cnt->conf.substream_port = 0;` in here as identified in the recently closed issue.

**Comment by Mr-Dave** [comment]
Couple of comments:

- The `scale_half_yuv420p` function does not use the subheight/subwidth so this will now error out on the developer warnings of unused variables.  It doesn't look like they are used at the calling point either so we may want to consider removing them from both the function parameters and calling routine.

- The only pix format that is currently used inside Motion is the yuv420.  This means that the `else` conditions on grey aren't really needed.  There are lots of those conditions left everywhere in the code so it could be addressed later in a general clean up that is needed everywhere. 

- The allocation of the substream image within the scaling function and freeing in stream module seems unusual and we may want to review it.  While it works, we may want to consider having that in the main Motion context like the other image allocations so we only allocate it once upon startup rather than at every frame.

- There is one line comment on the resetting of the substream port to zero when the init fails.

- I still have a bit more review to do on the scaling of the u/v components.

Overall, looks good.  Given the number of commits, I will be looking to do a squash merge.

**Comment by tosiara** [comment]
Thanks for the input. I will take a look and push requested changes to polish the PR

**Comment by tosiara** [comment]
Pushed few changes to address @Mr-Dave comments:
1. Disable substream if failed (to be in sync with https://github.com/Motion-Project/motion/pull/482)
2. Removed redundant parameters subwidth and subheight

**Pull request merged**